### PR TITLE
Validate pegout requests using pegout ids

### PR DESCRIPTION
**Pull request by scottmillner**
for now just doing an exact match on pegout amount which likely be refactored in the future to account for fees

### CODE REVIEW

File: crates/consensus/authority/src/frost_task.rs [hunk 1]
```diff
@@ -184,6 +206,58 @@ where
...
+                .map_err(|e| {
+                    error!(target: "consensus::authority::frost_task::validate_psbt_by_ids", "Failed to parse burn event {:?}", e);
+                    PsbtValidationError::FailedToValidatePsbtByIds
+                })?
+                .ok_or_else(|| {
+                    error!(target: "consensus::authority::frost_task::validate_psbt_by_ids", "Failed to get pegout data from burn event");
+                    PsbtValidationError::FailedToValidatePsbtByIds
+                })?;
+
+            // check if a corresponding output exists in the psbt
```
Comment (scottmillner) [inline crates/consensus/authority/src/frost_task.rs]: @0xBEEFCAF3 is this check sufficient?

im just looping through and finding an output that matches the pegout data
i don't have any way of knowing which output the pegout is, correct?

as in, we don't have a mapping of output index to pegout id, correct?

File: crates/consensus/authority/src/frost_task.rs [hunk 2]
```diff
@@ -184,6 +206,58 @@ where
...
+    pub(crate) async fn validate_psbt_by_ids(&self, psbt: Psbt) -> Result<(), PsbtValidationError> {
+        let pegout_ids = psbt
+            .outputs
+            .iter()
+            .filter_map(|output| match output.pegout_id() {
+                Some(pegout_id) => PegoutId::from_bytes(pegout_id.as_slice()).ok(),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+
```
Comment (arminsabouri) [inline crates/consensus/authority/src/frost_task.rs]: if len of pegout_ids is 0 something is off. Lets Error out here

File: crates/consensus/authority/src/frost_task.rs (hunk 1 above)
Comment (arminsabouri) [reply crates/consensus/authority/src/frost_task.rs]: Yes this looks correct to me

File: crates/consensus/authority/src/frost_task.rs [hunk 3]
```diff
@@ -184,6 +206,58 @@ where
...
+                    error!(target: "consensus::authority::frost_task::validate_psbt_by_ids", "Failed to parse burn event {:?}", e);
+                    PsbtValidationError::FailedToValidatePsbtByIds
+                })?
+                .ok_or_else(|| {
+                    error!(target: "consensus::authority::frost_task::validate_psbt_by_ids", "Failed to get pegout data from burn event");
+                    PsbtValidationError::FailedToValidatePsbtByIds
+                })?;
+
+            // check if a corresponding output exists in the psbt
+            if let Ok(transaction) = psbt.clone().extract_tx() {
```
Comment (arminsabouri) [inline crates/consensus/authority/src/frost_task.rs]: Nit: match here do remain consistent with the rest of this file's convention

File: crates/consensus/authority/src/frost_task.rs (hunk 2 above)
Comment (scottmillner) [reply crates/consensus/authority/src/frost_task.rs]: done!

File: crates/consensus/authority/src/frost_task.rs (hunk 3 above)
Comment (scottmillner) [reply crates/consensus/authority/src/frost_task.rs]: done!

File: crates/consensus/authority/src/frost_task.rs [hunk 4]
```diff
@@ -57,6 +68,12 @@ pub(crate) struct FrostNotification {
     pub(crate) psbt: Vec<u8>,
 }
 
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum PsbtValidationError {
+    #[error("Failed to validate psbt by ids")]
+    FailedToValidatePsbtByIds,
```
Comment (arminsabouri) [inline crates/consensus/authority/src/frost_task.rs]: Maybe we add a string here `(FailedToValidatePsbtByIds(String))` so we can seperate out which cases we are error'ing out

File: crates/consensus/authority/src/frost_task.rs [hunk 5]
```diff
@@ -184,6 +206,72 @@ where
...
+                    error!(target: "consensus::authority::frost_task::validate_psbt_by_ids", "Failed to get pegout data from burn event");
+                    PsbtValidationError::FailedToValidatePsbtByIds
+                })?;
+
+            // check if a corresponding output exists in the psbt
+            match psbt.clone().extract_tx() {
+                Ok(transaction) => {
+                    match transaction.output.iter().find(|output| {
+                        output.script_pubkey == destination.script_pubkey() &&
+                            output.value == amount
```
Comment (arminsabouri) [inline crates/consensus/authority/src/frost_task.rs]: Can you please leave a TODO here that strict amount checking will need to account for the fees coming out of the pegout outputs in the future

**Comment by arminsabouri** [comment]
@scottmillner how hard would it be to mock the db here and unit test this? This should definetly recieve some test coverage

File: crates/consensus/authority/src/frost_task.rs (hunk 4 above)
Comment (scottmillner) [reply crates/consensus/authority/src/frost_task.rs]: done!

File: crates/consensus/authority/src/frost_task.rs (hunk 5 above)
Comment (scottmillner) [reply crates/consensus/authority/src/frost_task.rs]: done!

**Review by arminsabouri — APPROVED** [review]

**Pull request merged**